### DISCUSSION TITLE
docs/filters

### DISCUSSION THREAD

**Discussion by giscus** [mikro-orm/mikro-orm#5439] [Comments]
# docs/filters

MikroORM has the ability to pre-define filter criteria and attach those filters to given entities. The application can then decide at runtime whether certain filters should be enabled and what their parameter values should be. Filters can be used like database views, but they are parameterized inside the application.

https://mikro-orm.io/docs/filters

**Comment by ramzini22** [comment]
Tell me please, how can i use filter (@Filter()) at queryBuilder

**Reply by B4nan** [reply]
You don't, QB usage needs to be explicit. Note that you don't need to use QB for vast majority of queries, using `em.find` can do most of what QB can.

What you can do is using `em.applyFilters()` to create a query based on filters (and feed that to `qb.where()`).

**Comment by awais33** [comment]
please help me
i am applying the filter in the nested many to many relations roles in this query in user.service.ts like this:

const users = await this.repo.findAndCount({roles: {slug: 'admin'}});

but i have get empty records, please tell me whats am i doing wrong?

this is my User model

import { Collection, Entity, EntityRepositoryType, Enum, Filter, ManyToMany, PrimaryKey, Property, Unique } from '@mikro-orm/core';
import { ObjectId } from '@mikro-orm/mongodb';
import { UserRepository } from './user.repository';
import { UserStatus } from 'src/constants/UserStatus';
import { UserType } from 'src/constants/UserType';
import { Role } from 'src/role/role.model';
import { MainEntity } from 'src/helper/db/base.entity';

@Entity({ repository: () => UserRepository })
export class User extends MainEntity {
  [EntityRepositoryType]?: UserRepository;

  @Property()
  firstName!: string;

  @Property()
  lastName!: string;

  @Property()
  password!: string;

  @Property()
  @Unique()
  email!: string;

  @Property({nullable: true})
  countryCode!: string;

  @Property({nullable: true})
  phone!: string;

  @Property()
  @Enum(() => UserType)
  userType!: string;

  @Property()
  @Enum(() => UserStatus)
  status!: UserStatus;

  @ManyToMany(() => Role, role => role.users, { owner: true, eager: true })
  roles = new Collection<Role>(this);
}

this is my Role model
import { Collection, Entity, EntityRepositoryType, Enum, ManyToMany, PrimaryKey, Property, Unique } from '@mikro-orm/core';
import { ObjectId } from '@mikro-orm/mongodb';
import { RoleRepository } from './role.repository';
import { User } from 'src/user/user.model';
import { RoleEnum } from 'src/constants/Role';
import { Exclude, Expose } from 'class-transformer';
import { Permission } from 'src/permission/permission.model';
import { MainEntity } from 'src/helper/db/base.entity';

@Entity({ repository: () => RoleRepository })
export class Role extends MainEntity {
  [EntityRepositoryType]?: RoleRepository;

  @Expose()
  @Property({ unique: true })
  @Enum(() => RoleEnum)
  role!: RoleEnum;

  @Expose()
  @Property({ unique: true })
  slug!: string;

  // @Exclude()
  @ManyToMany(() => User, user => user.roles)
  users = new Collection<User>(this);

  @ManyToMany(() => Permission, permission => permission.roles, { owner: true, eager: true})
  permissions = new Collection<Permission>(this);
}


this roleRepository

export class RoleRepository extends AppRepository<Role> {
    // custom methods
}

this is user repo
import { User } from "./user.model";
import { AppRepository } from "src/confugrations/database/mikro-orm/apprepository/app.repository";

export class UserRepository extends AppRepository<User> {
    // custom methods
}


this is common AppRepo

import { AnyEntity, EntityManager, EntityRepository, FilterQuery, ObjectId } from "@mikro-orm/mongodb";

export class AppRepository<T extends object> extends EntityRepository<T> {

    persist(entity: AnyEntity | AnyEntity[]): EntityManager {
        return this.em.persist(entity);
    }

    // async findAlll(entity: AnyEntity, filters: object, options: object): Promise<EntityManager> {
    //     return this.em.findByCursor(entity, filters, options);
    // }

    async persistAndFlush(entity: AnyEntity | AnyEntity[]): Promise<void> {
        await this.em.persistAndFlush(entity);
    }

    remove(entity: AnyEntity): EntityManager {
        return this.em.remove(entity);
    }

    async removeAndFlush(entity: AnyEntity): Promise<void> {
        await this.em.removeAndFlush(entity);
    }

    async flush(): Promise<void> {
        return this.em.flush();
    }

    // async findByIdOrFail(id: ObjectId): Promise<T> {
    //     const entity = await this.em.findOne(this.entityName, { _id: id } as FilterQuery<T>);
    //     if (!entity) {
    //         throw new Error(`Entity with ID ${id} not found`);
    //     }
    //     return entity;
    // }
}

**Comment by khangphung-vl** [comment]
Dear B4nan,
How can I extend the functionality of Filter to `getReference( )` ? 
Many of my work are done using `getReference`, and switching them all to `findOne` feels like some performance penalties to me (sorry if I'm wrong). But I highly doubt the possibility of this request.
Thanks for your help.

**Reply by B4nan** [reply]
Not entirely sure what you mean, maybe provide some code snippet. If you want `em.getReference()` to respect filters, that's a wontfix, all this method is doing is creating entity instance based on a PK, it doesn't fire any query - and filters are all handled on SQL level.

**Comment by Hanawa02** [comment]
How can I apply a filter based on 2 conditions? something like

```
@Filter({ name: 'library', cond: async (args) => {
  return {
    library: { id: args.library_id },
    collection: { library: { id: args.library_id }},
  };
} })
```

In this case I want the book to EITHER be part of this library, OR from a collection of the library

**Reply by B4nan** [reply]
I believe a simple `$or` should work as usual:

```ts
@Filter({ name: 'library', cond: async (args) => {
  return {
    $or: [
      { library: { id: args.library_id } },
      { collection: { library: { id: args.library_id } },
    ],
  };
} })
```
